components/net/subresource_integrity.rs

```diff
@@ -0,0 +1,164 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use net_traits::response::{Response, ResponseBody, ResponseType};
+use openssl::crypto::hash::{hash, Type as MessageDigest};
+use rustc_serialize::base64::{STANDARD, ToBase64};
+use std::collections::HashMap;
+use std::sync::MutexGuard;
+lazy_static! {
+    //Key is supported algorithm and value is priority
+    static ref SUPPORTED_ALGORITHM: HashMap<&'static str, usize> = {
+         let mut map = HashMap::new();
+         map.insert("sha256", 1);
+         map.insert("sha384", 2);
+         map.insert("sha512", 3);
+         map
+    };
+}
+
+#[derive(Clone)]
+pub struct SriEntry {
+    pub alg: String,
+    pub val: String,
+    // TODO : Current version of spec does not define any option.
+    // Can be refactored into appropriate datastructure when future
+    // spec has more details.
+    pub opt: Option<String>,
+}
+
+impl SriEntry {
+    pub fn new(alg: &str, val: &str, opt: Option<String>) -> SriEntry {
+        SriEntry {
+            alg: alg.to_owned(),
+            val: val.to_owned(),
+            opt: opt,
+        }
+    }
+}
+
+/// https://w3c.github.io/webappsec-subresource-integrity/#parse-metadata
+pub fn parsed_metadata(integrity_metadata: &str) -> Vec<SriEntry> {
+    // Step 1
+    let mut result = vec![];
+
+    // Step 3
+    let tokens: Vec<&str> = integrity_metadata.split(" ").collect();
+
+    for token in tokens {
+        let parsed_data: Vec<&str> = token.split("-").collect();
+
+        if parsed_data.len() > 0 {
+            let alg = parsed_data[0];
+
+            if SUPPORTED_ALGORITHM.contains_key(alg) {
+                let data: Vec<&str> = parsed_data[1].split("?").collect();
+                let digest = data[0];
+
+                let opt = if data.len() > 1 {
+                    Some(data[1].to_owned())
+                } else {
+                    None
+                };
+
+                result.push(SriEntry::new(alg, digest, opt))
+            }
+        }
+    }
+
+    return result;
+}
+
+/// https://w3c.github.io/webappsec-subresource-integrity/#getprioritizedhashfunction
+pub fn get_prioritized_hash_function(hash_func_left: &str, hash_func_right: &str) -> String {
```

Comment:
Let's return `Option<&str>` here instead, which will avoid the magic empty string behaviour.